File: services/api-go/internal/entity/task.go
// Package entity is the domain model. It imports nothing from this module and performs no I/O:
// its rules are the ones every transport and every store must agree on. The architecture test in
// internal/architecture_test.go fails the build when that stops being true.
package entity

import (
	"errors"
	"strings"
	"time"
	"unicode/utf8"
)

// Status is where a task is in its lifecycle.
type Status string

// The statuses a task moves through.
const (
	StatusTodo       Status = "todo"
	StatusInProgress Status = "in_progress"
	StatusDone       Status = "done"
)

// MaxTitleLength bounds a title, counted in characters rather than bytes.
const MaxTitleLength = 200

// Domain errors. Transports map them to their own error shapes; nothing here knows which
// transport is asking.
var (
	ErrEmptyTitle        = errors.New("title must not be empty")
	ErrTitleTooLong      = errors.New("title must be at most 200 characters")
	ErrUnknownStatus     = errors.New("status must be one of todo, in_progress, done")
	ErrInvalidTransition = errors.New("status transition not allowed")
	ErrNotFound          = errors.New("task not found")
)

// Task is a unit of work.
type Task struct {
	ID        string
	Title     string
	Status    Status
	CreatedAt time.Time
	UpdatedAt time.Time
}

// NewTask validates title and returns a task in StatusTodo. The caller supplies the ID and the
// time, which keeps this package deterministic.
func NewTask(id, title string, now time.Time) (Task, error) {
	title = strings.TrimSpace(title)
	if title == "" {
		return Task{}, ErrEmptyTitle
	}

	if utf8.RuneCountInString(title) > MaxTitleLength {
		return Task{}, ErrTitleTooLong
	}

	return Task{ID: id, Title: title, Status: StatusTodo, CreatedAt: now, UpdatedAt: now}, nil
}

// ParseStatus converts external input into a Status.
func ParseStatus(s string) (Status, error) {
	switch status := Status(s); status {
	case StatusTodo, StatusInProgress, StatusDone:
		return status, nil
	default:
		return "", ErrUnknownStatus
	}
}

// Transition returns a copy of t moved to next, or ErrInvalidTransition.
func (t Task) Transition(next Status, now time.Time) (Task, error) {
	if !canTransition(t.Status, next) {
		return Task{}, ErrInvalidTransition
	}

	t.Status = next
	t.UpdatedAt = now

	return t, nil
}

// canTransition lists every legal move. Anything absent is refused, including staying put.
func canTransition(from, to Status) bool {
	switch from {
	case StatusTodo:
		return to == StatusInProgress
	case StatusInProgress:
		return to == StatusTodo || to == StatusDone
	default:
		return false
	}
}

File: services/mcp-server/src/domain/task.ts
/**
 * The domain model: pure functions over plain data. No I/O, no clock, no randomness, and no imports
 * from outside this directory — not even a validation library, which is why `parseTask` is written
 * out. scripts/check-boundaries.mjs fails the build otherwise.
 *
 * The status rules are the same rules api-go and api-ts enforce. They are repeated here rather than
 * imported across modules ([ADR-0004](../../../../docs/adr/0004-independent-modules.md)) so this
 * service can be deleted, or kept alone, without touching anything else. What they buy: a tool can
 * tell a model which moves are legal instead of letting it discover a 409 by trying.
 */

export const STATUSES = ["todo", "in_progress", "done"] as const;
export type Status = (typeof STATUSES)[number];

export interface Task {
  readonly id: string;
  readonly title: string;
  readonly status: Status;
  readonly createdAt: string;
  readonly updatedAt: string;
}

export const MAX_TITLE_LENGTH = 200;

export type DomainErrorCode = "EMPTY_TITLE" | "TITLE_TOO_LONG" | "UNKNOWN_STATUS" | "INVALID_TRANSITION" | "NOT_FOUND" | "MALFORMED_TASK";

export class DomainError extends Error {
  readonly code: DomainErrorCode;

  constructor(code: DomainErrorCode, message: string) {
    super(message);
    this.name = "DomainError";
    this.code = code;
  }
}

/** Every legal move. Anything absent is refused, including staying put. */
const NEXT: Readonly<Record<Status, readonly Status[]>> = {
  todo: ["in_progress"],
  in_progress: ["todo", "done"],
  done: [],
};

export const nextStatuses = (status: Status): readonly Status[] => NEXT[status];

const isStatus = (value: unknown): value is Status => STATUSES.some((status) => status === value);

export function parseStatus(value: unknown): Status {
  if (isStatus(value)) return value;
  throw new DomainError("UNKNOWN_STATUS", `status must be one of ${STATUSES.join(", ")}`);
}

/** Refuses a title the API would refuse, so a bad one costs no request and gets a usable message. */
export function validateTitle(title: string): string {
  const trimmed = title.trim();
  if (trimmed === "") throw new DomainError("EMPTY_TITLE", "title must not be empty");
  // Counted in code points, not UTF-16 units, so the limit means the same as in api-go.
  if ([...trimmed].length > MAX_TITLE_LENGTH) {
    throw new DomainError("TITLE_TOO_LONG", `title must be at most ${MAX_TITLE_LENGTH} characters`);
  }
  return trimmed;
}

export function checkTransition(from: Status, to: Status): void {
  if (!NEXT[from].includes(to)) {
    const legal = NEXT[from].length === 0 ? "none" : NEXT[from].join(", ");
    throw new DomainError("INVALID_TRANSITION", `cannot move a task from ${from} to ${to}; legal moves from ${from}: ${legal}`);
  }
}

const isRecord = (value: unknown): value is Record<string, unknown> => typeof value === "object" && value !== null && !Array.isArray(value);

/**
 * A task as it arrives from the API, validated before anything else reads it. A service that trusts
 * a response it did not check reports the remote system's bugs as its own.
 */
export function parseTask(value: unknown): Task {
  if (!isRecord(value)) throw new DomainError("MALFORMED_TASK", "expected a task object");
  const { id, title, status, createdAt, updatedAt } = value;
  for (const [field, raw] of Object.entries({ id, title, createdAt, updatedAt })) {
    if (typeof raw !== "string" || raw === "") throw new DomainError("MALFORMED_TASK", `task.${field} must be a non-empty string`);
  }
  return {
    id: id as string,
    title: title as string,
    status: parseStatus(status),
    createdAt: createdAt as string,
    updatedAt: updatedAt as string,
  };
}

export const parseTasks = (value: unknown): readonly Task[] => {
  if (!Array.isArray(value)) throw new DomainError("MALFORMED_TASK", "expected an array of tasks");
  return value.map(parseTask);
};
